qdisk: refactor qdisk_start()

alltilla opened the pull request:
This PR refactors the monster `qdisk_start()` function. There are some minor edge case fixes, but no functional change is expected.

No NEWS file needed.

Signed-off-by: Attila Szakacs <szakacs.attila96@gmail.com>

kira-syslogng commented:
Build FAILURE

MrAnno commented on modules/diskq/qdisk.c:
```diff
@@ -1432,6 +1432,13 @@ _load_qdisk_file(QDisk *self, const gchar *filename, GQueue *qout, GQueue *qback
 }
 
+static gboolean
+_set_qdisk_file_defaults(QDisk *self)
```
I think this is not a good name.
Since I can't suggest a better one, I believe `_create_header()` should be kept in `qdisk_start()`.
Creating the header is a huge side effect of this function, I think this function should not exist.

alltilla replied:
As agreed IRL, renamed it to `_init_qdisk_file()`.

MrAnno commented on modules/diskq/qdisk.c:
```diff
@@ -1021,37 +1021,37 @@ _save_queue(QDisk *self, GQueue *q, QDiskQueuePosition *q_pos)
+      msg_info("Reliable disk-buffer state saved",
+               evt_tag_str("filename", self->filename),
+               evt_tag_long("qdisk_length", self->hdr->length));
+      return TRUE;
```
Zeroing out `q*_pos` is probably a useful operation here, because changing disk-buffer type is somewhat possible.

alltilla replied:
Nice catch, removed that commit.

MrAnno commented on modules/diskq/qdisk.c:
```diff
@@ -1037,8 +1037,6 @@ qdisk_save_state(QDisk *self, GQueue *qout, GQueue *qbacklog, GQueue *qoverflow)
         return FALSE;
     }
 
-  memcpy(self->hdr->magic, self->file_id, 4);
```
I think we should keep this here for now. Changing disk-buffer type is somewhat possible.
However, I think we could keep the magic setting in `_create_header()` too.

alltilla replied:
Nice catch! Fixed.

alltilla commented:
Rebased to master and fixed the comments.

---

Diff:
<details>

```
diff --git a/modules/diskq/qdisk.c b/modules/diskq/qdisk.c
index 5edaf1e13..d2c43fc87 100644
--- a/modules/diskq/qdisk.c
+++ b/modules/diskq/qdisk.c
@@ -1021,37 +1021,39 @@ error:
 gboolean
 qdisk_save_state(QDisk *self, GQueue *qout, GQueue *qbacklog, GQueue *qoverflow)
 {
-  if (self->options->reliable)
-    {
-      msg_info("Reliable disk-buffer state saved",
-               evt_tag_str("filename", self->filename),
-               evt_tag_long("qdisk_length", self->hdr->length));
-      return TRUE;
-    }
-
   QDiskQueuePosition qout_pos = { 0 };
   QDiskQueuePosition qbacklog_pos = { 0 };
   QDiskQueuePosition qoverflow_pos = { 0 };
 
-  qout_pos.count = qout->length / 2;
-  qbacklog_pos.count = qbacklog->length / 2;
-  qoverflow_pos.count = qoverflow->length / 2;
+  if (!self->options->reliable)
+    {
+      qout_pos.count = qout->length / 2;
+      qbacklog_pos.count = qbacklog->length / 2;
+      qoverflow_pos.count = qoverflow->length / 2;
 
-  if (!_save_queue(self, qout, &qout_pos) ||
-      !_save_queue(self, qbacklog, &qbacklog_pos) ||
-      !_save_queue(self, qoverflow, &qoverflow_pos))
-    return FALSE;
+      if (!_save_queue(self, qout, &qout_pos) ||
+          !_save_queue(self, qbacklog, &qbacklog_pos) ||
+          !_save_queue(self, qoverflow, &qoverflow_pos))
+        return FALSE;
+    }
+
+  memcpy(self->hdr->magic, self->file_id, 4);
 
   self->hdr->qout_pos = qout_pos;
   self->hdr->qbacklog_pos = qbacklog_pos;
   self->hdr->qoverflow_pos= qoverflow_pos;
 
-  msg_info("Disk-buffer state saved",
-           evt_tag_str("filename", self->filename),
-           evt_tag_long("qout_length", qout_pos.count),
-           evt_tag_long("qbacklog_length", qbacklog_pos.count),
-           evt_tag_long("qoverflow_length", qoverflow_pos.count),
-           evt_tag_long("qdisk_length", self->hdr->length));
+  if (!self->options->reliable)
+    msg_info("Disk-buffer state saved",
+             evt_tag_str("filename", self->filename),
+             evt_tag_long("qout_length", qout_pos.count),
+             evt_tag_long("qbacklog_length", qbacklog_pos.count),
+             evt_tag_long("qoverflow_length", qoverflow_pos.count),
+             evt_tag_long("qdisk_length", self->hdr->length));
+  else
+    msg_info("Reliable disk-buffer state saved",
+             evt_tag_str("filename", self->filename),
+             evt_tag_long("qdisk_length", self->hdr->length));
 
   return TRUE;
 }
@@ -1074,6 +1076,9 @@ _close_file(QDisk *self)
       close(self->fd);
       self->fd = -1;
     }
+
+  g_free(self->filename);
+  self->filename = NULL;
 }
 
 static void
@@ -1141,8 +1146,8 @@ _open_file(QDisk *self, const gchar *filename)
 static gboolean
 _create_file(QDisk *self, const gchar *filename)
 {
-  if (self->options->read_only)
-    return FALSE;
+  g_assert(!self->options->read_only);
+  g_assert(filename);
 
   if (self->options->disk_buf_size == -1)
     {
@@ -1170,8 +1175,11 @@ _create_file(QDisk *self, const gchar *filename)
   self->fd = fd;
   self->filename = g_strdup(filename);
 
-  if (self->options->prealloc)
-    return _preallocate(self, self->options->disk_buf_size);
+  if (self->options->prealloc && !_preallocate(self, self->options->disk_buf_size))
+    {
+      _close_file(self);
+      return FALSE;
+    }
 
   return TRUE;
 }
@@ -1192,7 +1200,12 @@ _create_header(QDisk *self)
   QDiskFileHeader nulled_hdr;
   memset(&nulled_hdr, 0, sizeof(nulled_hdr));
   if (!pwrite_strict(self->fd, &nulled_hdr, sizeof(nulled_hdr), 0))
-    return FALSE;
+    {
+      msg_error("Error occurred while allocating the header for a new queue file",
+                evt_tag_str("filename", self->filename),
+                evt_tag_error("error"));
+      return FALSE;
+    }
 
   memcpy(self->hdr->magic, self->file_id, 4);
 
@@ -1388,26 +1401,18 @@ _autodetect_disk_buf_size(QDisk *self)
 }
 
 static gboolean
-_handle_option_changes(QDisk *self)
+_ensure_disk_buf_size(QDisk *self)
 {
-  if (self->hdr->disk_buf_size != -1)
-    {
-      if (self->options->disk_buf_size != -1 && self->hdr->disk_buf_size != self->options->disk_buf_size)
-        {
-          msg_warning("WARNING: disk-buf-size() has changed since the last syslog-ng run. syslog-ng currently does "
-                      "not support changing the disk-buf-size() of existing disk-queues. Continuing with the old one",
-                      evt_tag_str("filename", self->filename),
-                      evt_tag_long("active_old_disk_buf_size", self->hdr->disk_buf_size),
-                      evt_tag_long("ignored_new_disk_buf_size", self->options->disk_buf_size));
-        }
-      return TRUE;
-    }
+  if (self->hdr->disk_buf_size == -1 && !_autodetect_disk_buf_size(self))
+    return FALSE;
 
-  if (!_autodetect_disk_buf_size(self))
+  if (self->hdr->disk_buf_size != self->options->disk_buf_size)
     {
-      msg_error("Failed to load disk-buf-size from disk-queue, and it was not set explicitly",
-                evt_tag_str("filename", self->filename));
-      return FALSE;
+      msg_warning("WARNING: disk-buf-size() has changed since the last syslog-ng run. syslog-ng currently does "
+                  "not support changing the disk-buf-size() of existing disk-queues. Continuing with the old one",
+                  evt_tag_str("filename", self->filename),
+                  evt_tag_long("active_old_disk_buf_size", self->hdr->disk_buf_size),
+                  evt_tag_long("ignored_new_disk_buf_size", self->options->disk_buf_size));
     }
 
   return TRUE;
@@ -1422,7 +1427,7 @@ _load_qdisk_file(QDisk *self, const gchar *filename, GQueue *qout, GQueue *qback
   if (!_load_state(self, qout, qbacklog, qoverflow))
     goto error;
 
-  if (!_handle_option_changes(self))
+  if (!_ensure_disk_buf_size(self))
     goto error;
 
   return TRUE;
@@ -1433,10 +1438,15 @@ error:
 }
 
 static gboolean
-_set_qdisk_file_defaults(QDisk *self)
+_init_qdisk_file(QDisk *self)
 {
-  self->file_size = QDISK_RESERVED_SPACE;
-  return _create_header(self);
+  if (!_create_header(self))
+    return FALSE;
+
+  if (!self->file_size)
+    self->file_size = QDISK_RESERVED_SPACE;
+
+  return TRUE;
 }
 
 static gboolean
@@ -1445,7 +1455,7 @@ _create_qdisk_file(QDisk *self, const gchar *filename, GQueue *qout, GQueue *qba
   if (!_create_file(self, filename))
     goto error;
 
-  if (!_set_qdisk_file_defaults(self))
+  if (!_init_qdisk_file(self))
     goto error;
 
   return TRUE;
@@ -1489,9 +1499,6 @@ qdisk_init_instance(QDisk *self, DiskQueueOptions *options, const gchar *file_id
 void
 qdisk_stop(QDisk *self)
 {
-  g_free(self->filename);
-  self->filename = NULL;
-
   _close_file(self);
 }
 
diff --git a/modules/diskq/tests/test_qdisk.c b/modules/diskq/tests/test_qdisk.c
index 14a731ae9..638ae3772 100644
--- a/modules/diskq/tests/test_qdisk.c
+++ b/modules/diskq/tests/test_qdisk.c
@@ -29,6 +29,7 @@
 #include "scratch-buffers.h"
 
 #include <unistd.h>
+#include <sys/stat.h>
 
 /* QDisk-internal: the frame is a 4-byte integer */
 #define FRAME_LENGTH 4
@@ -354,6 +355,27 @@ Test(qdisk, completely_full_and_then_emptied_qdisk_should_update_positions_prope
   cleanup_qdisk(filename, qdisk);
 }
 
+Test(qdisk, prealloc)
+{
+  const gchar *filename = "test_prealloc.rqf";
+
+  DiskQueueOptions *opts = construct_diskq_options(TDISKQ_RELIABLE, MIN_DISK_BUF_SIZE);
+  disk_queue_options_set_prealloc(opts, TRUE);
+  QDisk *qdisk = qdisk_new(opts, "TEST");
+
+  qdisk_start(qdisk, filename, NULL, NULL, NULL);
+
+  struct stat file_stats;
+  cr_assert(stat(filename, &file_stats) == 0, "Stat call failed, errno: %d", errno);
+  gint64 real_size = file_stats.st_size;
+
+  cr_assert_eq(qdisk_get_file_size(qdisk), MIN_DISK_BUF_SIZE);
+  cr_assert_eq(qdisk_get_file_size(qdisk), real_size);
+
+  qdisk_stop(qdisk);
+  cleanup_qdisk(filename, qdisk);
+}
+
 static void
 setup(void)
 {
```

</details>

alltilla commented on modules/diskq/qdisk.c:
```diff
@@ -1265,9 +1311,6 @@ qdisk_start(QDisk *self, const gchar *filename, GQueue *qout, GQueue *qbacklog,
   g_assert(!qdisk_started(self));
 
-  if (self->options->read_only && !filename)
-    return FALSE;
```
Thanks, added it to `_create_file()`

alltilla commented on modules/diskq/qdisk.c:
```diff
@@ -1346,56 +1410,14 @@ qdisk_start(QDisk *self, const gchar *filename, GQueue *qout, GQueue *qbacklog,
         {
-          msg_error("Error occurred while initializing the new queue file",
-                    evt_tag_str("filename", self->filename),
-                    evt_tag_error("error"));
```
Thanks, I added it back with a bit more explicit message.

alltilla commented on modules/diskq/qdisk.c:
```diff
@@ -1388,6 +1387,32 @@ _autodetect_disk_buf_size(QDisk *self)
 }
 
+static gboolean
+_handle_option_changes(QDisk *self)
```
I renamed the function and reordered it a bit:
```c
static gboolean
_ensure_disk_buf_size(QDisk *self)
{
  if (self->hdr->disk_buf_size == -1 && !_autodetect_disk_buf_size(self))
    return FALSE;

  if (self->hdr->disk_buf_size != self->options->disk_buf_size)
    {
      msg_warning("WARNING: disk-buf-size() has changed since the last syslog-ng run. syslog-ng currently does "
                  "not support changing the disk-buf-size() of existing disk-queues. Continuing with the old one",
                  evt_tag_str("filename", self->filename),
                  evt_tag_long("active_old_disk_buf_size", self->hdr->disk_buf_size),
                  evt_tag_long("ignored_new_disk_buf_size", self->options->disk_buf_size));
    }

  return TRUE;
}
```

MrAnno replied:
Thanks.

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

bazsi commented on modules/diskq/qdisk.c:
```diff
@@ -1201,6 +1201,8 @@ _create_header(QDisk *self)
       return FALSE;
     }
 
+  memcpy(self->hdr->magic, self->file_id, 4);
```
sizeof() perhaps?

alltilla replied:
Thanks, fixed in both cases.

bazsi commented on modules/diskq/qdisk.c:
```diff
@@ -1440,8 +1440,13 @@ _load_qdisk_file(QDisk *self, const gchar *filename, GQueue *qout, GQueue *qback
+  if (!self->file_size)
+    self->file_size = QDISK_RESERVED_SPACE;
+
+  return TRUE;
```
I would appreciate some a more elaborate comment on why this fixes file_size after prealloc.

alltilla replied:
Good point, I have added details to the commit:
```
    qdisk: fix setting file_size after prealloc

    Setting file_size must be done after each operation, which can alter the
    file's size. Creating the header on the disk is an operation like that,
    but we cannot blindly set the file_size to the header's size, because
    we might have preallocated the file before that which could have already
    set file_size.

    This commit fixes this issue by checking if file_size is already set or
    not, and only setting it to the header's size, when necessary.
```

bazsi reviewed (approved):
I went through all commits and they are quite easy to follow and the final form of qdisk_start() looks awesome.

There was only a single commit I didn't grasp from the commit summary, where I'd appreciate a more elaborate commit message, but otherwise this looks good to me.

alltilla commented:
@bazsi Thank you for taking the time to review this! :)

Rebased to master and fixed the review comments.

alltilla commented:
@kira-syslogng do stresstest

kira-syslogng commented:
Kira-stress-test: Build SUCCESS

MrAnno commented on modules/diskq/qdisk.c:
```diff
@@ -1393,6 +1392,24 @@ _autodetect_disk_buf_size(QDisk *self)
+_ensure_disk_buf_size(QDisk *self)
+{
+  if (self->hdr->disk_buf_size == -1 && !_autodetect_disk_buf_size(self))
+    return FALSE;
```
`Failed to load disk-buf-size from disk-queue, and it was not set explicitly` has disappeared. Is this intentional?

alltilla replied:
Yes, actually all of the fail branches of `_autodetect_disk_buf_size()` log an error, so I think there is no need to log another one on the call site.

MrAnno commented on modules/diskq/qdisk.c:
```diff
@@ -1440,8 +1440,13 @@ _load_qdisk_file(QDisk *self, const gchar *filename, GQueue *qout, GQueue *qback
+    return FALSE;
+
+  if (!self->file_size)
+    self->file_size = QDISK_RESERVED_SPACE;
```
This (and the need for a detailed explanation) suggests that the placement of preallocation is not the best from a clean-code point of view.

`_create_file()` has the preallocation side-effect, which is a bit far from `_init_qdisk_file()`: it's in `_create_qdisk_file`.

I think this should look something like this:

```
_create_qdisk_file():
  _create_file();
  _create_header();
  if (!_preallocate())
    file_size = ...;
```

or

```
_create_qdisk_file():
  _create_file();
 
  if (need_prealloc)
    _prealloc();
  
  _create_header();

  if(!need_prealloc)
    file_size = ...
```

Or anything similar. I think the point still is that `_init_qdisk_file()` should not exist.

MrAnno replied:
Let me know if you don't want to fix this. We can merge it as-is.

alltilla replied:
I agree, but if you don't mind I would do this in a new PR.

MrAnno replied:
Fixed in #4380.

MrAnno commented:
LGTM otherwise.
